model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java

```diff
@@ -139,15 +145,19 @@ public void addFederatedIdentity(RealmModel realm, UserModel user, FederatedIden
 
         getEntityById(realm, user.getId())
                 .ifPresent(userEntity ->
-                        userEntity.addFederatedIdentity(UserFederatedIdentityEntity.fromModel(socialLink)));
+                        userEntity.addFederatedIdentity(MapUserFederatedIdentityEntity.fromModel(socialLink)));
     }
 
     @Override
     public boolean removeFederatedIdentity(RealmModel realm, UserModel user, String socialProvider) {
         LOG.tracef("removeFederatedIdentity(%s, %s, %s)%s", realm, user.getId(), socialProvider, getShortStackTrace());
-        return getEntityById(realm, user.getId())
-                .map(entity -> entity.removeFederatedIdentity(socialProvider))
-                .orElse(false);
+
+        Optional<MapUserEntity> entityById = getEntityById(realm, user.getId());
+        if (!entityById.isPresent()) return false;
+
+        return entityById.map(u -> u.getFederatedIdentity(socialProvider))
+                .map(uc -> uc.isPresent() && entityById.get().removeFederatedIdentity(uc.get()))
+                .orElse(true); // TODO: make removeFederatedIdentity return Boolean so the caller can correctly handle "I don't know" null answer
```

Comment:
Does this have to change? Could `removeFederatedIdentity(String federatedIdentity)` be retained and the logic moved to the `AbstractUserEntity` implementation?
It seems to me that in the `AbstractUserEntity` it could be as simple as using `removeIf(fi -> Objects.equals(fi.getIdentityProvider(), federatedIdentity)` while keeping existing provider implementation.

If so, similarly this would happen in other places in this file

Reply:
This is a good idea, I will rewrite it.